### PR TITLE
feat: txpool block building fallback

### PR DESCRIPTION
**Pull request by mattsse**
prep for #3702

this prepares for local block building for pending block if no block from CL is available yet.

the actually impl will be in a followup because I need to refactor/reuse some stuff from the basic payload builder impl

### CODE REVIEW

File: crates/rpc/rpc/src/eth/api/block.rs [hunk 1]
```diff
@@ -121,7 +124,12 @@ where
 
         if block_id.is_pending() {
             // Pending block can be fetched directly without need for caching
-            return Ok(self.provider().pending_block()?)
+            let maybe_pending = self.provider().pending_block()?;
+            return if maybe_pending.is_some() {
+                return Ok(maybe_pending)
+            } else {
+                self.local_pending_block().await
+            }
```
Comment (gakonst) [inline crates/rpc/rpc/src/eth/api/block.rs:132]: very nice

File: crates/rpc/rpc/src/eth/api/mod.rs [hunk 2]
```diff
@@ -201,6 +207,74 @@ where
...
+                if block_env.number.to::<u64>() == pending.block.number &&
+                    pending.block.parent_hash == parent_header.parent_hash &&
+                    now <= pending.expires_at
+                {
+                    return Ok(Some(pending.block.clone()))
+                }
+            }
+
+            // TODO(mattsse): actually build the pending block
+            Ok(None)
```
Comment (gakonst) [inline crates/rpc/rpc/src/eth/api/mod.rs:272]: 👍

File: crates/rpc/rpc/src/eth/api/mod.rs [hunk 3]
```diff
@@ -201,6 +207,74 @@ where
...
+
+    /// Returns the locally built pending block
+    pub(crate) async fn local_pending_block(&self) -> EthResult<Option<SealedBlock>> {
+        let pending = self.pending_block_env_and_cfg()?;
+        if pending.origin.is_actual_pending() {
+            return Ok(pending.origin.into_actual_pending())
+        }
+
+        // no pending block from the CL yet, so we need to build it ourselves via txpool
+        self.on_blocking_task(|this| async move {
```
Comment (gakonst) [inline crates/rpc/rpc/src/eth/api/mod.rs:254]: Makes me think we should probably add docs on which ops are blocking and when we recommend spawning blocking?

**Review by gakonst — APPROVED** [review]
supportive

File: crates/rpc/rpc/src/eth/api/mod.rs (hunk 3 above)
Comment (mattsse) [reply crates/rpc/rpc/src/eth/api/mod.rs:254]: good point, will do!

**Pull request merged**